crates/ibc-elc-client-cw/src/entrypoint.rs

```
@@ -0,0 +1,156 @@
+        ctx.set_checksum(msg.checksum);
+
+        let any_client_state = Any::decode(&mut msg.client_state.as_slice())?;
+        let any_consensus_state = Any::decode(&mut msg.consensus_state.as_slice())?;
+
+        let res = lc.create_client(&ctx, any_client_state.clone(), any_consensus_state.clone())?;
+
+        ctx.store_client_state(res.height, any_client_state)?;
+        ctx.store_consensus_state(res.height, any_consensus_state)?;
+        ctx.store_update_meta(res.height, ctx.host_timestamp().into(), ctx.host_height())?;
+
+        Ok(Response::default().set_data(to_json_binary(&ContractResult::success())?))
+    }
+
+    fn sudo(
+        lc: &Self::LightClient,
+        deps: DepsMut<'_>,
+        env: Env,
+        msg: SudoMsg,
+    ) -> Result<Response, ContractError> {
+        let mut ctx = Context::new_mut(deps, env);
+
+        let result = match msg {
+            SudoMsg::UpdateState(msg) => {
+                let any_message = Any::decode(msg.client_message.as_slice())?;
+                let res = match lc.update_client(&ctx, ctx.client_id().clone(), any_message)? {
+                    UpdateClientResult::UpdateState(d) => d,
+                    _ => panic!("unexpected non-UpdateState client message"),
+                };
+
+                ctx.store_client_state(res.height, res.new_any_client_state)?;
+                ctx.store_consensus_state(res.height, res.new_any_consensus_state)?;
+                ctx.store_update_meta(res.height, ctx.host_timestamp().into(), ctx.host_height())?;
+
+                ContractResult::success().heights(vec![res.height])
+            }
+            SudoMsg::UpdateStateOnMisbehaviour(msg) => {
+                let any_message = Any::decode(msg.client_message.as_slice())?;
+                let res = match lc.update_client(&ctx, ctx.client_id().clone(), any_message)? {
+                    UpdateClientResult::Misbehaviour(d) => d,
+                    _ => panic!("unexpected non-Misbehaviour client message"),
```

If this panic can occur due to the tx submitted by the relayer, we should return an error instead.

---

`ibc-go` calls `CheckForMisbehaviour` at first, and then calls `UpdateStateOnMisbehaviour` only if the former function returns true.
For this crate, `CheckForMisbehaviour` is also implemented based on `lc.update_client`.
Therefore in `UpdateStateOnMisbehaviour`, `lc.update_client` always returns `UpdateClientResult::Misbehaviour`.